File: src/storage/section-codec.js
/**
 * Section content codec for the `on-demand` (compacted) storage profile.
 *
 * `storage compact` zstd-compresses document_sections.content_text /
 * content_json in place to shrink an ultra-lean install. Compression is
 * opportunistic — a value is only stored compressed when that actually saves
 * bytes, so small sections stay plain strings. Decoding is therefore
 * type-directed and fully backward-compatible: an uncompressed corpus reads
 * through untouched.
 *   - a string column value is returned as-is,
 *   - a BLOB (a Uint8Array, read back from a compressed cell) is zstd-decoded.
 */

const ZSTD_MAGIC = [0x28, 0xb5, 0x2f, 0xfd]

/**
 * Compress a section content value for storage. Returns the original string
 * when compression wouldn't save space (so tiny rows aren't bloated), `''` for
 * empty, or null for null/undefined.
 * @param {string|null|undefined} text
 * @returns {string|Uint8Array|null}
 */
export function encodeSectionContent(text) {
  if (text == null) return null
  const str = String(text)
  if (str.length === 0) return ''
  const raw = Buffer.from(str, 'utf8')
  const z = Bun.zstdCompressSync(raw)
  return z.length < raw.length ? z : str
}

/**
 * Decode a section content value read from the DB. Strings pass through;
 * zstd BLOBs are inflated to a UTF-8 string.
 * @param {string|Uint8Array|null|undefined} value
 * @returns {string|null|undefined}
 */
export function decodeSectionContent(value) {
  if (value == null || typeof value === 'string') return value
  if (
    value.length >= 4 &&
    value[0] === ZSTD_MAGIC[0] && value[1] === ZSTD_MAGIC[1] &&
    value[2] === ZSTD_MAGIC[2] && value[3] === ZSTD_MAGIC[3]
  ) {
    return Buffer.from(Bun.zstdDecompressSync(value)).toString('utf8')
  }
  // Non-zstd blob (shouldn't happen) — best-effort UTF-8 decode.
  return Buffer.from(value).toString('utf8')
}

/**
 * In-place decode of `content_text` / `content_json` on a raw section row.
 * No-op on an uncompressed corpus.
 * @template {{ content_text?: unknown, content_json?: unknown }} T
 * @param {T} row
 * @returns {T}
 */
export function decodeSectionRow(row) {
  if (!row) return row
  if (row.content_text != null) row.content_text = decodeSectionContent(row.content_text)
  if (row.content_json != null) row.content_json = decodeSectionContent(row.content_json)
  return row
}

File: swift/Sources/ADWrite/SectionCodec.swift
// The section content codec — the native port of `src/storage/section-codec.js`
// (the `compact` storage profile's cell format). `storage compact` zstd-compresses
// `document_sections.content_text` / `content_json` in place; compression is
// opportunistic (a value is stored compressed only when that actually saves
// bytes, so small sections stay plain strings), which makes decoding fully
// type-directed and backward-compatible: TEXT passes through, a BLOB is a zstd
// frame (magic 28 b5 2f fd) inflated — or plain UTF-8 bytes as a best-effort
// fallback, exactly like `decodeSectionContent`.
//
// The encoder compresses at zstd level 3 — the library default
// (`ZSTD_CLEVEL_DEFAULT`), which is what `Bun.zstdCompressSync` uses when the
// JS codec omits the level. The exact frame bytes may still differ from a
// Bun-compacted corpus (different zstd builds/parameters), which is fine: the
// format contract is the type split (TEXT vs zstd BLOB), not the frame bytes.

import ADArchive
public import ADStorage

/// Encode/decode `document_sections` content cells (section-codec.js).
public enum SectionCodec {
    /// `encodeSectionContent(text)`: nil → NULL, empty → `''`, else the zstd
    /// frame when it is strictly smaller than the UTF-8 bytes, else the string.
    public static func encode(_ text: String?) -> SQLiteValue {
        guard let text else { return .null }
        if text.isEmpty { return .text("") }
        let raw = Array(text.utf8)
        if let compressed = ZstdEncoder.compress(raw, level: 3), compressed.count < raw.count {
            return .blob(compressed)
        }
        return .text(text)
    }

    /// `decodeSectionContent(value)`: TEXT passes through; a BLOB with the zstd
    /// magic is inflated; any other BLOB decodes as UTF-8 bytes; NULL → nil.
    public static func decodeText(_ value: SQLiteValue?) -> String? {
        switch value {
            case .text(let text):
                return text
            case .blob(let bytes):
                if bytes.count >= 4, bytes[0] == 0x28, bytes[1] == 0xB5, bytes[2] == 0x2F, bytes[3] == 0xFD,
                    let inflated = ZstdDecoder.decompress(bytes)
                {
                    return String(decoding: inflated, as: UTF8.self)
                }
                return String(decoding: bytes, as: UTF8.self)
            default:
                return nil
        }
    }
}
